UserService_DbContext

KostinP opened the pull request:
UserService project is added to the solution.
DbContext is defined and the database is created locally using PM console.

KostinP committed:
Base of UserService

We created an empty ASP.NET Core Web Application and added dependencies that you will need to migrate the database

KostinP committed:
We added entities for DB

We created four entities (Achievement, Company, Position, User) based on models sheme.

We used two attributes: [Key] and [NotMapped]
Key is for the primary key definition.
NotMapped key is needed so as not to duplicate data in different tables, since this data already lies in other tables.
* Certificates lies in FileService.
* Positions lies in table Position.
* Achievements lies in table Achievements.

Created UserServiceDbContext where we added our entities.

KostinP committed:
We created migrations to DB

We added "ConnectionStrings" to appsettings.json and pointed it in Startup to establish connection with the SQL Server.

And entered the command for PM:
-- AddMigration InitialCreate
-- Update-Database  -Migration InitialCreate

And the database was created successfully

mekk1t reviewed (commented):
As @DenisKrivoruchko11 pointed out, we should add a new folder "Services" and add all future projects there. I think we discussed this when we were talking about the tests -- we will have the separate folder called "Tests", so logically we should also have "Services" folder.
So our solution structure will be like:
Solution/Services/UserService
Solution/Services/ProjectService
etc.

Other than that, everything looks cool.

ifelseelif commented on DigitalOffice/UserService/Database/Entities/Company.cs:
```diff
@@ -0,0 +1,13 @@
+        [Key]
+        public Guid Id { get; set; }
+        public string Name { get; set; }
+        public Guid CEO { get; set; }
```
I think CEOId should be used here

mekk1t replied:
Actually yeah. Since we mark all guids as PropertyId...
Agree.

ifelseelif commented on DigitalOffice/UserService/Properties/launchSettings.json:
```diff
@@ -0,0 +1,27 @@
+    "UserService": {
+      "commandName": "Project",
+      "launchBrowser": true,
+      "applicationUrl": "https://localhost:5001;http://localhost:5000",
```
I think it's necessary to change ports, because when launching several services errors may occur

mekk1t replied:
A great idea. We'll make adjustments to the ProjectService as well

denis-kriv reviewed (commented):
we have two files .sln in this branch or is it a github bug?

KostinP committed:
Minor fixes

* Deleted second .sln file.
* UserService moved to folder "Services".
* Updated ports.
* .UseHttpsRedirection() is included  in the Startup.
* Corrected field names

lalagvana committed:
Automated Migrations

Migrations are applied to the addressed DbContext in a new private static method that gets a scope of the dbContext and updates its database.

nmamro commented on DigitalOffice/Services/UserService/Migrations/20200711045501_InitialCreate.cs:
```diff
@@ -0,0 +1,88 @@
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    Name = table.Column<string>(nullable: true),
```
Is it ok that there're some questionable nullables? For example, company name, user emal, first/last name etc.

lalagvana committed:
Entity Update

We added entities for many-to-many relationships:
*Relationship between User and Position.
Position or User shouldn't include information about Start and End Time for position itself, because we will store extra data there(too much not very different entries). So we moved these fields(StartTime, EndTime)  to relationship entity.
*Relationship between User and File(Certificate).
*Relationship between User and Achievement.
Achievement or User shouldn't include information about Time of getting an achievement itself, so it must be in their relationship entity, just like in UserPosition model. And we moved this field(Time) to relationship model.

mekk1t commented on DigitalOffice/Services/UserService/Database/Entities/UserAchievement.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class UserAchievement
+    {
+        public Guid UserId { get; set; }
+        public virtual User User { get; set; }
+        public Guid AchievementId { get; set; }
+        public virtual Achievement Achievement { get; set; }
```
Don't forget that such navigational properties should also be checked in Achievement entity and User entity.

Since this works as a link between users and achievements, please, look up into many-to-many relationship in EF Core.

Anyway, after you've changed Entities definition, migrate the changes and EF Core will tell if everything's correct :)

KostinP committed:
Update DigitalOffice.sln

Added a solution folder "Services" into .sln file

KostinP committed:
Updated DbContext, added migration

Set up the relationship between the table.

KostinP committed:
Set up required attribute and nullable types and added to migration

denis-kriv reviewed (commented):
Resolve conflicts please

djino97 commented on DigitalOffice/Services/UserService/Database/Entities/User.cs:
```diff
@@ -0,0 +1,31 @@
+        [Required]
+        public string PasswordHash { get; set; }
+        public Guid? AvatarFileId { get; set; }
+        [NotMapped]
+        public ICollection<UserCertificateFile> CertificatesFilesIds { get; set; }
+        [NotMapped]
+        public ICollection<UserPosition> PositionsIds { get; set; }
+        [NotMapped]
+        public ICollection<UserAchievement> AchievementsIds { get; set; }
```
Why didn't you remove atribute NotMapped? If it stay in property, your columns will not be created.

nmamro replied:
> tay in property, your columns will no

Do we want collection type columns?

djino97 replied:
We use collection type, because ICollection<T> used for normal properties, just because its a well-known interface with minimal overheads, i.e. it is very slightly quicker to create an ICollection than a IList. Also,  if NotMapped stay in property, just tables will not create relationships between themselves.

KostinP committed:
Removed NotMapped attribute

New migration is not needed, as it will be empty.

KostinP committed:
Updated from develop

nmamro commented on DigitalOffice/Services/UserService/UserService.csproj:
```diff
@@ -0,0 +1,16 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.5" />
```
Isn't 3.1.6 the latest stable one?

KostinP replied:
Packages updated

KostinP committed:
Updated packages
